src/test/ui/polymorphization/closure_in_upvar/fn.rs

```diff
@@ -0,0 +1,29 @@
+// build-pass
+// compile-flags:-Zpolymorphize=on -Zsymbol-mangling-version=v0
+
+fn foo(f: impl Fn()) {
+    let x = |_: ()| ();
+
+    // Don't use `f` in `y`, but refer to `x` so that the closure substs contain a reference to
+    // `x` that will differ for each instantiation despite polymorphisation of the varying
+    // argument.
+    let y = || x(());
```

Comment:
What happens if `y` considers `f` used and `x` doesn't?

Reply:
If `y` considers `f` used then it’ll produce two mono items with different symbols (because `f` varies) - with this PR, in both mono items, the upvar subst for `x` will be polymorphised. `x` would still produce a single mono item like now.

However, it occurs to me that we might still have a issue if `x` considers `f` used. In that case, this patch wouldn’t change the upvar substitutions in `y` for `x` and therefore we’d still produce two mono items that aren’t equal but produce the same symbol. Not sure how to deal with that case - don’t consider upvars in hashing and equality?

Reply:
Can `y` consider `f` unused if it uses `x` which considers it used?

Reply:
> However, it occurs to me that we might still have a issue if x considers f used. In that case, this patch wouldn’t change the upvar substitutions in y for x and therefore we’d still produce two mono items that aren’t equal but produce the same symbol. Not sure how to deal with that case - don’t consider upvars in hashing and equality?

After discussion with @lcnr and testing locally, can confirm that we don't have an issue here.

> Can y consider f unused if it uses x which considers it used?

If we allowed this then I think we could run into issues with the symbol mangling again - because the upvar of `x` in `y` would differ depending on the instantiation of `f`, which would result in two mono items which have the same symbol, triggering `assert_symbols_are_distinct` with the new mangling scheme.

Reply:
I've added tests for the cases we've discussed here.